urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/json/parser/ConfigurableJSONBuilder.java

```diff
@@ -0,0 +1,307 @@
+	}
+    }
+    
+    @Override
+    protected void fireArrayEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireObjectBegin() {
+	JSONObject obj = null;
+	if( this.currentMemberName != null )
+	    obj = this.factory.createObject( this.currentMemberName );
+	else if( this.currentArrayIndex != null )
+	    obj = this.factory.createObject( this.currentArrayIndex );
+	else
+	    obj = this.factory.createObject();
+
+	this.valueStack.push( obj ); // new JSONObject( new TreeMap<String,JSONValue>() ) );
+	this.memberNameStack.push( this.currentMemberName );
+	this.arrayIndexStack.push( this.currentArrayIndex );
+    }
+
+    @Override
+    protected void fireObjectEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireMemberBegin() {
+	// NOOP
+    }
+
+    @Override
+    protected void fireMemberNameRead( String token ) {
+	this.currentMemberName = token;
+    }
+
+    @Override
+    protected void fireMemberEnd() {
+	try {
+	    this.valueStack.peek().getObject().put( this.currentMemberName, 
+						    this.currentMemberValue
+						    );
+	} catch( JSONException e ) {
+	    throw new RuntimeException( "Fatal error: unexpected JSONException caught when trying to retrieve JSON object.", e );
+	}	
+    }
+
+    @Override
+    protected void fireNumberRead( String number ) {
+	
+	try {
+
+	    // May throw NumberFormatException
+	    Number num =  JSONNumber.parseNumber(number);
+
+	    if( this.currentMemberName != null )
+		currentMemberValue = this.factory.createNumber( num, this.currentMemberName );
+	    else if( this.currentArrayIndex != null )
+		this.currentMemberValue = this.factory.createNumber( num, this.currentArrayIndex );
+	    else
+		this.currentMemberValue = this.factory.createNumber( num );
+
+	} catch( NumberFormatException e ) {
+	    // Store as a string??? Throw exception??? 
+	    // Actually this means there is a type error in the parser, so this excption should NEVER occur
+	    throw new RuntimeException( "Ooops, this number is invalid: " + number );
+	    //this.currentMemberValue = new JSONString( number );
+	}
+	
+    }
+
+    @Override
+    protected void fireStringRead( String token ) {
+	if( this.currentMemberName != null )
+	    this.currentMemberValue = this.factory.createString( token, this.currentMemberName );
+	else if( this.currentArrayIndex != null )
+	    this.currentMemberValue = this.factory.createString( token, this.currentArrayIndex );
+	else
+	    this.currentMemberValue = this.factory.createString( token );
+    }
+    
+    @Override
+    protected void fireTrueRead( String value ) {
+	if( this.currentMemberName != null )
+	    currentMemberValue = this.factory.createBoolean( new Boolean(true), this.currentMemberName );
+	else if( this.currentArrayIndex != null )
+	    this.currentMemberValue = this.factory.createBoolean( new Boolean(true), this.currentArrayIndex );
+	else
+	    this.currentMemberValue = this.factory.createBoolean( new Boolean(true) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 236:241, 247:252.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.